src/lsm/tree.zig

```diff
@@ -865,10 +870,10 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             assert(tree.compaction_io_pending == 0);
             assert(tree.compaction_callback == null);
 
-            // Avoid checkpointing if this is not the last beat in the compaction measure.
+            // Assert that this is the last beat in the compaction measure.
             const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
             const last_beat_in_measure = config.lsm_batch_multiple - 1;
-            if (compaction_beat != last_beat_in_measure) return callback(tree);
+            assert(last_beat_in_measure == compaction_beat);
```

Comment:
@kprotty would you double-review this please?

Reply:
:+1: Just for extra context:

- The replica only calls `checkpoint()` at the correct tick (if it doesn't, that's a bug).
- I think this is the safer approach because otherwise the replica might call `checkpoint()`, resume from the callback, and then assume (incorrectly) that the LSM actually checkpointed.

Reply:
100%